### DISCUSSION TITLE
When is healthcheck 200 and how can I make sure migrations and seed script has run?

### DISCUSSION THREAD

**Discussion by einarpersson** [pocketbase/pocketbase#3813] [Q&A]
In my CI pipeline I would like to know when the database is completely ready, eg. for running e2e-tests.

This requires that 
- migrations have run
- seed script has run

```js
onAfterBootstrap(({ app }) => {
	const ENV = $os.getenv('ENV')

	console.log(`Running seed.pb.js/onAfterBootstrap in <${ENV}> env...`)

	if (ENV === 'DEV' || ENV === 'PULL_REQUEST') {
		app.onBeforeServe().add(() => {
			console.log('Seeding database...')

			// Seed
			// ...

```

1. What does health 200 really means? 
2. Any ideas on how the database can tell the world that it is ready, including arbitrary stuff like seeding?
3. I don't know if it makes sense, but could there be some kind of toggle that I can switched on, or configure a condition for the health check? Or some metadata? (I see that the healtcheck returns `"data": { "canBackup": ... } `, could this be customized? Or if there was a hook for the health check request that could modified? I don't know what the best solution is, I just want to share my issue.

I guess I could do something of a workaround like checking for the existence of a test user or something, which also could be configured as a custom route, but it feels a bit hacky. 

(Right now I have a type-generation step in CI which is flaky but it always works on the second manual re-run. It runs in github workflow against a fly io machine)

**Comment by ganigeorgiev** [comment]
> 1. What does health 200 really means?

The HTTP server (and respectively the `GET /api/health` endpoint) should be accessible AFTER the app bootstrap process and all new migrations have been applied.

About 2) and 3) - I don't understand what do you mean. The above code is expected to be executed BEFORE the health checkpoint is accessible. The response at the moment is not customizible.

I think the problem relies in your type-generation step as probably is executed in a separate process and runs concurrently. If that's the case you can try to execute the script via the [`$os.cmd()`](https://pocketbase.io/jsvm/functions/_os.cmd.html) in the `onBeforeServe` handler (`onBeforeServe` is fired right before starting the HTTP server):

```js
// prepare the command to execute
const cmd = $os.cmd('ls', '-sl')

// execute the command and return its standard output as string
const output = String.fromCharCode(...cmd.output());
```

The above is expected to be blocking as it will run as part of the same process.

**Reply by ganigeorgiev** [reply]
_Also note that you don't need to wrap `onBeforeServe` in `onAfterBootstrap` as the first one is is expected to be executed after the second one._

**Reply by ganigeorgiev** [reply]
On second read, I misunderstood how you run your code.

> It runs in github workflow against a fly io machine

Then this will be always racy. But note the health endpoint is expected to be accessible AFTER the app initialization, so if you are asking "how to know when the db will be ready?" - when the health check returns 200.

If that doesn't help, please elaborate and provide a more detailed code sample.

**Reply by einarpersson** [reply]
So healthcheck returns 200 after migrations have been successfully run, got it. ✅ 

However, I also want to run a seed script. But now that I think of it, I may simply rethink where / how I seed it. Currently I have done it as part of bootstrap (as you saw) but perhaps I should create a custom endpoint or something, for both seeding and resetting. Or  just use the `js-sdk` for this. But I think I added it to the bootstrap for convenience, so that it always runs when the db is restarted.

Well, I think I got some new ideas. Thank you for your time (once again)
